// redux-first-history/build/es5/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createReduxHistoryContext = exports.reachify = exports.LOCATION_CHANGE = exports.CALL_HISTORY_METHOD = exports.goForward = exports.goBack = exports.go = exports.replace = exports.push = void 0;
var actions_1 = require("./actions");
Object.defineProperty(exports, "push", { enumerable: true, get: function () { return actions_1.push; } });
Object.defineProperty(exports, "replace", { enumerable: true, get: function () { return actions_1.replace; } });
Object.defineProperty(exports, "go", { enumerable: true, get: function () { return actions_1.go; } });
Object.defineProperty(exports, "goBack", { enumerable: true, get: function () { return actions_1.goBack; } });
Object.defineProperty(exports, "goForward", { enumerable: true, get: function () { return actions_1.goForward; } });
Object.defineProperty(exports, "CALL_HISTORY_METHOD", { enumerable: true, get: function () { return actions_1.CALL_HISTORY_METHOD; } });
Object.defineProperty(exports, "LOCATION_CHANGE", { enumerable: true, get: function () { return actions_1.LOCATION_CHANGE; } });
var reachify_1 = require("./reachify");
Object.defineProperty(exports, "reachify", { enumerable: true, get: function () { return reachify_1.reachify; } });
var create_1 = require("./create");
Object.defineProperty(exports, "createReduxHistoryContext", { enumerable: true, get: function () { return create_1.createReduxHistoryContext; } });
